CI: TestSpireDelegateClient_GetCertificateForIdentity/get_certificate_for_numeric_identity

```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃   FAIL  package: github.com/cilium/cilium/pkg/auth/spire   ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

--- FAIL: TestSpireDelegateClient_GetCertificateForIdentity (0.02s)
--- FAIL: TestSpireDelegateClient_GetCertificateForIdentity/get_certificate_for_numeric_identity (0.00s)

    certificate_provider_test.go:393: SpireDelegateClient.GetCertificateForIdentity() = &{[[48 130 1 221 48 130 1 70 160 3 2 1 2 2 1 1 48 13 6 9 42 134 72 134 247 13 1 1 11 5 0 48 0 48 32 24 15 48 48 48 49 48 49 48 49 48 48 48 48 48 48 90 23 13 50 53 48 54 49 57 49 51 48 57 48 51 90 48 0 48 129 159 48 13 6 9 42 134 72 134 247 13 1 1 1 5 0 3 129 141 0 48 129 137 2 129 129 0 203 198 50 36 248 151 4 0 9 245 142 231 166 87 93 52 98 227 182 149 113 221 128 246 99 60 178 41 147 146 224 12 14 140 114 8 39 30 87 33 195 49 171 11 136 110 176 124 58 210 236 48 96 246 219 2 10 82 8 80 88 100 22 94 90 152 56 234 37 107 209 35 0 49 248 43 27 146 163 1 166 134 222 61 99 55 13 191 196 131 16 172 134 128 238 100 162 21 132 156 109 39 52 8 93 126 28 162 132 193 85 95 193 139 219 57 181 123 192 228 147 119 28 87 213 184 37 207 2 3 1 0 1 163 101 48 99 48 14 6 3 85 29 15 1 1 255 4 4 3 2 7 128 48 29 6 3 85 29 37 4 22 48 20 6 8 43 6 1 5 5 7 3 1 6 8 43 6 1 5 5 7 3 2 48 50 6 3 85 29 17 1 1 255 4 40 48 38 134 36 115 112 105 102 102 101 58 47 47 115 112 105 102 102 101 46 99 105 108 105 117 109 47 105 100 101 110 116 105 116 121 47 49 50 51 52 48 13 6 9 42 134 72 134 247 13 1 1 11 5 0 3 129 129 0 139 2 249 196 117 104 184 249 113 142 249 185 22 184 171 66 246 230 83 52 151 35 215 193 98 18 251 50 157 150 76 168 9 7 100 231 51 25 37 187 176 35 129 74 118 176 62 223 158 88 12 35 239 0 127 250 244 107 161 125 100 213 144 18 167 6 196 152 128 112 42 227 32 107 214 101 57 191 79 165 244 137 123 242 171 50 101 169 222 75 206 225 4 22 63 204 169 85 52 102 157 61 17 163 183 135 72 154 66 131 207 87 236 140 108 226 1 210 154 215 64 192 143 243 117 141 45 218]] 0x40003b5ab0 [] [] [] 0x40003fac08}, want &{[[48 130 1 221 48 130 1 70 160 3 2 1 2 2 1 1 48 13 6 9 42 134 72 134 247 13 1 1 11 5 0 48 0 48 32 24 15 48 48 48 49 48 49 48 49 48 48 48 48 48 48 90 23 13 50 53 48 54 49 57 49 51 48 57 48 51 90 48 0 48 129 159 48 13 6 9 42 134 72 134 247 13 1 1 1 5 0 3 129 141 0 48 129 137 2 129 129 0 203 198 50 36 248 151 4 0 9 245 142 231 166 87 93 52 98 227 182 149 113 221 128 246 99 60 178 41 147 146 224 12 14 140 114 8 39 30 87 33 195 49 171 11 136 110 176 124 58 210 236 48 96 246 219 2 10 82 8 80 88 100 22 94 90 152 56 234 37 107 209 35 0 49 248 43 27 146 163 1 166 134 222 61 99 55 13 191 196 131 16 172 134 128 238 100 162 21 132 156 109 39 52 8 93 126 28 162 132 193 85 95 193 139 219 57 181 123 192 228 147 119 28 87 213 184 37 207 2 3 1 0 1 163 101 48 99 48 14 6 3 85 29 15 1 1 255 4 4 3 2 7 128 48 29 6 3 85 29 37 4 22 48 20 6 8 43 6 1 5 5 7 3 1 6 8 43 6 1 5 5 7 3 2 48 50 6 3 85 29 17 1 1 255 4 40 48 38 134 36 115 112 105 102 102 101 58 47 47 115 112 105 102 102 101 46 99 105 108 105 117 109 47 105 100 101 110 116 105 116 121 47 49 50 51 52 48 13 6 9 42 134 72 134 247 13 1 1 11 5 0 3 129 129 0 139 2 249 196 117 104 184 249 113 142 249 185 22 184 171 66 246 230 83 52 151 35 215 193 98 18 251 50 157 150 76 168 9 7 100 231 51 25 37 187 176 35 129 74 118 176 62 223 158 88 12 35 239 0 127 250 244 107 161 125 100 213 144 18 167 6 196 152 128 112 42 227 32 107 214 101 57 191 79 165 244 137 123 242 171 50 101 169 222 75 206 225 4 22 63 204 169 85 52 102 157 61 17 163 183 135 72 154 66 131 207 87 236 140 108 226 1 210 154 215 64 192 143 243 117 141 45 218]] 0x40003b5960 [] [] [] 0x40003fa608}
```

Full logs: https://github.com/cilium/cilium/actions/runs/15757273749/job/44415390343

Seen on #40023

/cc @cilium/sig-servicemesh

--- a/pkg/auth/spire/certificate_provider_test.go
+++ b/pkg/auth/spire/certificate_provider_test.go
@@ -11,7 +11,6 @@ import (
 	"crypto/x509/pkix"
 	"math/big"
 	"net/url"
-	"reflect"
 	"testing"
 	"time"
 
@@ -50,9 +49,8 @@ func TestSpireDelegateClient_NumericIdentityToSNI(t *testing.T) {
 				},
 				log: hivetest.Logger(t).With(logfields.LogSubsys, "spire"),
 			}
-			if got := s.NumericIdentityToSNI(tt.args.id); got != tt.want {
-				t.Errorf("SpireDelegateClient.NumericIdentityToSNI() = %v, want %v", got, tt.want)
-			}
+			got := s.NumericIdentityToSNI(tt.args.id)
+			assert.Equal(t, tt.want, got)
 		})
 	}
 }
@@ -101,13 +99,8 @@ func TestSpireDelegateClient_SNIToNumericIdentity(t *testing.T) {
 				log: hivetest.Logger(t).With(logfields.LogSubsys, "spire"),
 			}
 			got, err := s.SNIToNumericIdentity(tt.args.sni)
-			if (err != nil) != tt.wantErr {
-				t.Errorf("SpireDelegateClient.SNIToNumericIdentity() error = %v, wantErr %v", err, tt.wantErr)
-				return
-			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("SpireDelegateClient.SNIToNumericIdentity() = %v, want %v", got, tt.want)
-			}
+			assert.Equal(t, tt.wantErr, err != nil)
+			assert.Equal(t, tt.want, got)
 		})
 	}
 }
@@ -137,9 +130,8 @@ func TestSpireDelegateClient_sniToSPIFFEID(t *testing.T) {
 				},
 				log: hivetest.Logger(t).With(logfields.LogSubsys, "spire"),
 			}
-			if got := s.sniToSPIFFEID(tt.args.id); got != tt.want {
-				t.Errorf("SpireDelegateClient.sniToSPIFFEID() = %v, want %v", got, tt.want)
-			}
+			got := s.sniToSPIFFEID(tt.args.id)
+			assert.Equal(t, tt.want, got)
 		})
 	}
 }
@@ -210,13 +202,8 @@ func TestSpireDelegateClient_ValidateIdentity(t *testing.T) {
 				log: hivetest.Logger(t).With(logfields.LogSubsys, "spire"),
 			}
 			got, err := s.ValidateIdentity(tt.args.id, tt.args.cert)
-			if (err != nil) != tt.wantErr {
-				t.Errorf("SpireDelegateClient.ValidateIdentity() error = %v, wantErr %v", err, tt.wantErr)
-				return
-			}
-			if got != tt.want {
-				t.Errorf("SpireDelegateClient.ValidateIdentity() = %v, want %v", got, tt.want)
-			}
+			assert.Equal(t, tt.wantErr, err != nil)
+			assert.Equal(t, tt.want, got)
 		})
 	}
 }
@@ -259,13 +246,8 @@ func TestSpireDelegateClient_GetTrustBundle(t *testing.T) {
 				trustBundle: tt.trustBundle,
 			}
 			got, err := s.GetTrustBundle()
-			if (err != nil) != tt.wantErr {
-				t.Errorf("SpireDelegateClient.GetTrustBundle() error = %v, wantErr %v", err, tt.wantErr)
-				return
-			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("SpireDelegateClient.GetTrustBundle() = %v, want %v", got, tt.want)
-			}
+			assert.Equal(t, tt.wantErr, err != nil)
+			assert.Equal(t, tt.want, got)
 		})
 	}
 }
@@ -493,9 +475,7 @@ func TestSpireDelegateClient_SubscribeToRotatedIdentities(t *testing.T) {
 				break
 			}
 
-			if !reflect.DeepEqual(got, tt.events) {
-				t.Errorf("SpireDelegateClient.SubscribeToRotatedIdentities() = %v, want %v", got, tt.events)
-			}
+			assert.Equal(t, tt.events, got)
 		})
 	}
 }

--- a/pkg/auth/spire/certificate_provider_test.go
+++ b/pkg/auth/spire/certificate_provider_test.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cilium/hive/hivetest"
 	delegatedidentityv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/agent/delegatedidentity/v1"
 	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/stretchr/testify/assert"
 
 	"github.com/cilium/cilium/pkg/auth/certs"
 	"github.com/cilium/cilium/pkg/identity"
@@ -385,13 +386,25 @@ func TestSpireDelegateClient_GetCertificateForIdentity(t *testing.T) {
 				svidStore: svidStore,
 			}
 			got, err := s.GetCertificateForIdentity(tt.args.id)
-			if (err != nil) != tt.wantErr {
-				t.Errorf("SpireDelegateClient.GetCertificateForIdentity() error = %v, wantErr %v", err, tt.wantErr)
+			assert.Equal(t, tt.wantErr, err != nil)
+			if tt.want == nil {
+				assert.Nil(t, got)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("SpireDelegateClient.GetCertificateForIdentity() = %v, want %v", got, tt.want)
-			}
+			assert.Equal(t, tt.want.Certificate, got.Certificate)
+			// Standard library hides private fields in the
+			// PrivateKey for FIPS compliance, which causes
+			// "assert.Equal()" to report inconsistent results,
+			// leading to unreliable failures of the test. Since
+			// we know the types here we can cast and use the type
+			// specific Equals.
+			wantKey := tt.want.PrivateKey.(*rsa.PrivateKey)
+			gotKey := got.PrivateKey.(*rsa.PrivateKey)
+			assert.True(t, wantKey.Equal(gotKey),
+				"SpireDelegateClient.GetCertificateForIdentity().PrivateKey: %v\nwant: %v",
+				got.PrivateKey, tt.want.PrivateKey,
+			)
+			assert.Equal(t, tt.want.Leaf, got.Leaf)
 		})
 	}
 }

